sdks/java/io/sparkreceiver/src/main/java/org/apache/beam/sdk/io/sparkreceiver/ReadFromSparkReceiverWithOffsetDoFn.java

```diff
@@ -141,15 +145,42 @@ public boolean hasRecords() {
     @Override
     public void start(Receiver<V> sparkReceiver) {
       this.sparkReceiver = sparkReceiver;
-      try {
-        new WrappedSupervisor(
-            sparkReceiver,
-            new SparkConf(),
-            objects -> {
-              V record = (V) objects[0];
-              recordsQueue.offer(record);
+
+      final SerializableFunction<Object[], Void> storeFn =
+          (input) -> {
+            if (input == null) {
               return null;
-            });
+            }
+
+            if (input[0] instanceof ByteBuffer) {
+              final ByteBuffer byteBuffer = ((ByteBuffer) input[0]).asReadOnlyBuffer();
+              final byte[] bytes = new byte[byteBuffer.limit()];
+              byteBuffer.get(bytes);
+              final V record = SerializationUtils.deserialize(bytes);
+              recordsQueue.offer(record);
+            } else if (input[0] instanceof Iterator) {
+              final Iterator<V> iterator = (Iterator<V>) input[0];
+              while (iterator.hasNext()) {
+                V record = (V) iterator.next();
+                recordsQueue.offer(record);
+              }
+            } else if (input[0] instanceof ArrayBuffer) {
+              final ArrayBuffer<V> arrayBuffer = (ArrayBuffer<V>) input[0];
+              final Iterator<V> iterator = arrayBuffer.iterator();
+              while (iterator.hasNext()) {
```

Comment:
Please add unit tests to fully cover this logic.

Reply:
Added three new tests that cover this logic.